File: HoTT-MdlViewer-JavaFX/src/main/java/gde/mdl/ui/SelectFromMemory.java
/**
 * HoTT Transmitter Config Copyright (C) 2013 Oliver Treichel
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package gde.mdl.ui;

import gde.mdl.messages.Messages;
import gde.mdl.ui.background.GetModelDataTask;
import gde.mdl.ui.background.LoadFromMemoryService;
import gde.model.serial.ModelInfo;
import javafx.concurrent.Task;
import javafx.scene.control.ButtonBar.ButtonData;
import javafx.scene.control.ButtonType;
import javafx.scene.control.ListView;
import javafx.scene.control.SelectionMode;

/**
 * @author oli@treichels.de
 */
public class SelectFromMemory extends SelectFromTransmitter {
    private final ListView<String> listView = new ListView<>();
    private final LoadFromMemoryService service;

    public SelectFromMemory() {
        setTitle(Messages.getString("LoadFromMemory"));

        listView.getSelectionModel().setSelectionMode(SelectionMode.SINGLE);
        listView.setOnMouseClicked(e -> handleDoubleClick(e));

        borderPane.setCenter(listView);

        service = new LoadFromMemoryService(listView);
        service.portNameProperty().bind(comboBox.valueProperty());
    }

    @Override
    protected Task<Model> getResult(final ButtonType b) {
        if (b.getButtonData() == ButtonData.OK_DONE && hasResult()) {
            final int index = listView.getSelectionModel().getSelectedIndex();
            final ModelInfo info = service.getValue().get(index);
            final GetModelDataTask task = new GetModelDataTask(listView, comboBox.getValue(), info);

            task.start();
            return task;
        }

        return null;
    }

    @Override
    protected boolean hasResult() {
        return listView.getSelectionModel().getSelectedIndex() != -1;
    }
}

File: HoTT-MdlViewer-JavaFX/src/main/java/de/treichels/hott/mdlviewer/javafx/SelectFromMemory.kt
/**
 * HoTT Transmitter Config Copyright (C) 2013 Oliver Treichel
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program. If not, see <http:></http:>//www.gnu.org/licenses/>.
 */
package de.treichels.hott.mdlviewer.javafx

import de.treichels.decoder.HoTTSerialPort
import de.treichels.hott.model.enums.ModelType
import de.treichels.hott.model.serial.ModelInfo
import javafx.beans.binding.BooleanBinding
import javafx.concurrent.Task
import javafx.scene.control.ListView
import javafx.scene.control.SelectionMode
import tornadofx.*
import java.util.concurrent.Callable

/**
 * @author Oliver Treichel &lt;oli@treichels.de&gt;
 */
class SelectFromMemory : SelectFromTransmitter() {
    private var listView by singleAssign<ListView<String>>()
    private var modelInfoList: List<ModelInfo>? = null

    // dialog is ready when the user selects an item from the list
    override fun isReady(): BooleanBinding = listView.selectionModel.selectedItemProperty().isNotNull

    // refresh listView from Transmitter memory
    override fun refreshUITask() = listView.loadFromMemory()

    // load the selected item as Model
    override fun getResultCallable(): Callable<Model>? {
        val index = listView.selectionModel.selectedIndex
        val modelInfo = modelInfoList?.get(index)

        return if (modelInfo != null) {
            Callable {
                val modelData = serialPort?.use { p ->
                    p.open()
                    p.getModelData(modelInfo)
                }

                Model(modelInfo, modelData)
            }
        } else null
    }

    init {
        title = messages["select_from_memory_title"]

        // add to UI
        root.center {
            listView = listview {
                selectionModel.selectionMode = SelectionMode.SINGLE
                setOnMouseClicked { handleDoubleClick(it) }
                items.add(messages["loading"])
            }
        }
    }

    /**
     * Refresh a [ListView] with a list of all models from the transmitter memory using the [HoTTSerialPort] from parent.
     */
    @Synchronized
    private fun ListView<String>.loadFromMemory(): Task<*> {
        items.clear()

        // add temporary placeholder
        items.add(messages["loading"])

        // fill list in background
        return listView.runAsyncWithOverlay {
            serialPort?.use { p ->
                p.open()
                p.allModelInfos
            }?.filter { it?.modelType != ModelType.Unknown } ?: listOf()
        }.success { list ->
            // save result
            modelInfoList = list

            // remove temporary placeholder
            items.clear()

            // translate list of ModelInfo to list of String
            items.addAll(list.map { info -> String.format("%02d: %c%s.mdl", info.modelNumber + 1, info.modelType?.char, info.modelName) })
        }
    }
}

